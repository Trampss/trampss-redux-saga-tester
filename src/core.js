export default middlewares => saga => (...params) => (mocks) => {
  let transform = value => value

  if (middlewares) {
    const mockedMiddlewares = middlewares.map(middleware => middleware(mocks))

    transform = (value) => {
      let transformedValue = value
      for (let i = 0; i < mockedMiddlewares.length; i += 1) {
        transformedValue = mockedMiddlewares[i](transformedValue)
      }
      return transformedValue
    }
  }

  const gen = saga(...params)
  const values = []

  let next = { value: undefined, done: false }
  do {
    next = transform(gen.next(next.value))
    values.push(next)
  } while (!next.done)

  return values
}
