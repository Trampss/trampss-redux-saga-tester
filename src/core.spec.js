/* eslint-env jest */
import core from './core'

describe('core', () => {
  function* saga(param1, param2) {
    yield param1
    const variable = yield param2
    yield variable
  }

  describe('without middlewares/mocks', () => {
    const c = core()
    it('should returns all generator values', () => {
      expect(c(saga)('foo', 'bar')(/* no mocks */)).toMatchSnapshot()
    })
  })

  describe('with middleware', () => {
    const middleware1 = mocks => next => Object.assign({}, next, { test: mocks })
    const middleware2 = mocks => next => Object.assign({}, next, { test2: mocks })
    const c = core([middleware1, middleware2])

    it('should transform next values by middlewares', () => {
      expect(c(saga)('foo', 'bar')('yeah this is a simple mock')).toMatchSnapshot()
    })
  })
})
