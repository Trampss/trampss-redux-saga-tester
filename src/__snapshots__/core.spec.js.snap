// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core with middleware should transform next values by middlewares 1`] = `
Array [
  Object {
    "done": false,
    "test": "yeah this is a simple mock",
    "test2": "yeah this is a simple mock",
    "value": "foo",
  },
  Object {
    "done": false,
    "test": "yeah this is a simple mock",
    "test2": "yeah this is a simple mock",
    "value": "bar",
  },
  Object {
    "done": false,
    "test": "yeah this is a simple mock",
    "test2": "yeah this is a simple mock",
    "value": "bar",
  },
  Object {
    "done": true,
    "test": "yeah this is a simple mock",
    "test2": "yeah this is a simple mock",
    "value": undefined,
  },
]
`;

exports[`core without middlewares/mocks should returns all generator values 1`] = `
Array [
  Object {
    "done": false,
    "value": "foo",
  },
  Object {
    "done": false,
    "value": "bar",
  },
  Object {
    "done": false,
    "value": "bar",
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;
